File: .tmp_py4013/batch_11_dump/3225_maximum_score_from_grid_operations.js
// LeetCode 3225 - Maximum Score From Grid Operations
// https://leetcode.com/problems/maximum-score-from-grid-operations/

var maximumScore = function(grid) {
    const n = grid.length;
    const prefix = Array.from({length: n}, () => new Array(n + 1).fill(0));
    for (let j = 0; j < n; j++) {
        for (let i = 0; i < n; i++) prefix[j][i + 1] = prefix[j][i] + grid[i][j];
    }
    let prevPick = new Array(n + 1).fill(0), prevSkip = new Array(n + 1).fill(0);
    for (let j = 1; j < n; j++) {
        const currPick = new Array(n + 1).fill(0), currSkip = new Array(n + 1).fill(0);
        for (let curr = 0; curr <= n; curr++) {
            for (let prev = 0; prev <= n; prev++) {
                if (curr > prev) {
                    const score = prefix[j - 1][curr] - prefix[j - 1][prev];
                    currPick[curr] = Math.max(currPick[curr], prevSkip[prev] + score);
                    currSkip[curr] = Math.max(currSkip[curr], prevSkip[prev] + score);
                } else {
                    const score = prefix[j][prev] - prefix[j][curr];
                    currPick[curr] = Math.max(currPick[curr], prevPick[prev] + score);
                    currSkip[curr] = Math.max(currSkip[curr], prevPick[prev]);
                }
            }
        }
        prevPick = currPick;
        prevSkip = currSkip;
    }
    let ans = -Infinity;
    for (const v of prevPick) ans = Math.max(ans, v);
    return ans;
};

File: .tmp_rs4013/rust_check15/3225_maximum_score_from_grid_operations.rs
struct Solution;
// LeetCode 3225 - Maximum Score From Grid Operations
// https://leetcode.com/problems/maximum-score-from-grid-operations/

impl Solution {
    pub fn maximum_score(grid: Vec<Vec<i32>>) -> i64 {
        let n = grid.len();
        let mut prefix = vec![vec![0i64; n + 1]; n];
        for j in 0..n {
            for i in 0..n {
                prefix[j][i + 1] = prefix[j][i] + grid[i][j] as i64;
            }
        }
        let mut prev_pick = vec![0i64; n + 1];
        let mut prev_skip = vec![0i64; n + 1];
        for j in 1..n {
            let mut curr_pick = vec![0i64; n + 1];
            let mut curr_skip = vec![0i64; n + 1];
            for curr in 0..=n {
                for prev in 0..=n {
                    if curr > prev {
                        let score = prefix[j - 1][curr] - prefix[j - 1][prev];
                        curr_pick[curr] = curr_pick[curr].max(prev_skip[prev] + score);
                        curr_skip[curr] = curr_skip[curr].max(prev_skip[prev] + score);
                    } else {
                        let score = prefix[j][prev] - prefix[j][curr];
                        curr_pick[curr] = curr_pick[curr].max(prev_pick[prev] + score);
                        curr_skip[curr] = curr_skip[curr].max(prev_pick[prev]);
                    }
                }
            }
            prev_pick = curr_pick;
            prev_skip = curr_skip;
        }
        *prev_pick.iter().max().unwrap()
    }
}

fn main() {}
